mutt/date.c

```diff
@@ -447,6 +460,206 @@ uint64_t mutt_date_now_ms(void)
+ * and does not support negative numbers. Empty strings are parsed as zero.
+ */
+static int parse_small_uint(const char *str, const char *end, int *val)
+{
+  const char *ptr = str;
+  int v = 0;
+  while (ptr < end && ptr < str + 5 && *ptr >= '0' && *ptr <= '9')
+  {
+    v = v * 10 + (*ptr - '0');
+    ++ptr;
+  }
+  *val = v;
+  return ptr - str;
+}
+
+/**
+ * mutt_date_parse_rfc5322_strict - Parse a date string in RFC822 format,
+ *   without any comments or extra whitespace (except a comment at the
+ *   very end, since that is very common for time zones).
+ *
+ * This is a fairly straightforward implementation in the hope of extracting
+ * the valid cases quickly, i.e., without having to resort to a regex.
+ * The hard cases are left to a regex implementation further down in
+ * mutt_date_parse_date() (which calls us).
+ *
+ * Spec: https://tools.ietf.org/html/rfc5322#section-3.3
+ */
+static time_t mutt_date_parse_rfc5322_strict(const char *s, struct Tz *tz_out)
+{
+  size_t len = strlen(s);
+
+  /* Skip over the weekday, if any. */
+  if (len >= 5 && s[4] == ' ' &&
+      (eqi4(s, "Mon,") || eqi4(s, "Tue,") || eqi4(s, "Wed,") ||
+       eqi4(s, "Thu,") || eqi4(s, "Fri,") || eqi4(s, "Sat,") || eqi4(s, "Sun,")))
+  {
+    s += 5;
+    len -= 5;
+  }
+
+  while (len > 0 && *s == ' ')
+  {
+    ++s;
+    --len;
+  }
+
+  if (len == 0 || *s < '0' || *s > '9')
+    return -1;
+
+  struct tm tm = { 0 };
+
+  /* Day */
+  int mday_len = parse_small_uint(s, s + len, &tm.tm_mday);
+  if (mday_len == 0 || mday_len > 2 || tm.tm_mday > 31)
+    return -1;
+  s += mday_len;
+  len -= mday_len;
+
+  if (len == 0 || *s != ' ')
+    return -1;
+  ++s;
+  --len;
+
+  /* Month */
+  if (len < 3)
+    return -1;
+  tm.tm_mon = mutt_date_check_month(s);
+  if (tm.tm_mon == -1)
+    return -1;
+  s += 3;
+  len -= 3;
+
+  if (len == 0 || *s != ' ')
+    return -1;
+  ++s;
+  --len;
+
+  /* Year */
+  int year_len = parse_small_uint(s, s + len, &tm.tm_year);
+  if (year_len != 2 && year_len != 4)
+    return -1;
+  if (tm.tm_year < 50)
+    tm.tm_year += 100;
+  else if (tm.tm_year >= 1900)
+    tm.tm_year -= 1900;
+  s += year_len;
+  len -= year_len;
+
+  if (len == 0 || *s != ' ')
+    return -1;
+  ++s;
+  --len;
+
+  /* Hour */
+  if (len < 3 || s[0] < '0' || s[0] > '2' || s[1] < '0' || s[1] > '9' || s[2] != ':')
+    return -1;
+  tm.tm_hour = (s[0] - '0') * 10 + (s[1] - '0');
+  if (tm.tm_hour > 23)
+    return -1;
+  s += 3;
+  len -= 3;
+
+  /* Minute */
+  if (len < 2 || s[0] < '0' || s[0] > '5' || s[1] < '0' || s[1] > '9')
+    return -1;
+  tm.tm_min = (s[0] - '0') * 10 + (s[1] - '0');
+  if (tm.tm_min > 59)
+    return -1;
+  s += 2;
+  len -= 2;
+
+  /* Second (optional) */
+  if (len > 0 && s[0] == ':')
+  {
+    ++s;
+    --len;
+    if (len < 2 || s[0] < '0' || s[0] > '5' || s[1] < '0' || s[1] > '9')
+      return -1;
+    tm.tm_sec = (s[0] - '0') * 10 + (s[1] - '0');
+    if (tm.tm_sec > 60)
+      return -1;
```

Comment:
For min and sec, is there a way for the second comparison to fail if the first one succeeds? For sec, shouldn't the second one be `> 59` (unless you want to support leap seconds)?

Reply:
The existing code supports 60 (which I would assume comes from the RFC); I should probably fix the code to s[0] > '6' instead.

Reply:
Leap seconds are rare enough that checking for 59 and dropping to slow path if 60 is encountered seems pretty ok to me. Can't think of another situation where 60 is encountered at least.